src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs

```diff
@@ -233,14 +234,21 @@ public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver generat
             }
         }
 
-        internal sealed class ProjectCompilationOptionsAction(
-            ProjectState oldProjectState,
-            ProjectState newProjectState)
-            : TranslationAction(oldProjectState, newProjectState)
+        internal sealed class ProjectCompilationOptionsAction : TranslationAction
         {
+            public ProjectCompilationOptionsAction(
+                ProjectState oldProjectState,
+                ProjectState newProjectState) : base(oldProjectState, newProjectState)
+            {
+                if (NewProjectState.CompilationOptions == null)
+                {
+                    throw new NotSupportedException(WorkspacesResources.Removing_compilation_options_is_not_supported);
```

Comment:
trying to reconcile this with this line in the pr: https://github.com/dotnet/roslyn/pull/74289/files#diff-980cc9deba20d375eabfcefc6256ec0d4d58880e1952809e960d4fdb5f1e404bR667

Reply:
We construct a new state with `CompilationOptions` null and then see if the state changed (might have been null already). Only when the state actually changed we throw here.

We could detect the non-null -> null options change earlier but it would duplicate some code.

Reply:
basically, i'm confsued since part of the PR seems to be all about allowing null options, and this now seems to be about blowing up on them.

Reply:
The issue is null should be fine to set as long as it doesn't change from not-null. Say, you're calling WithProjectInfo on a TypeScript project that never had the options and the new ProjectInfo also doesn't have the options. We shouldn't blow up there.

Reply:
i would be ok with this explicitly being checked for at the highest level and bailing out if it is null->null  as a no-op.  Or null->not-null.  But not-null->null blowing up would make me feel a TON more confident on things.